src/wavm/mpi.cpp

```diff
@@ -843,6 +1202,202 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(env,
+                               I32 reorder,
+                               I32 newCommPtr)
+{
+    faabric::util::getLogger()->debug("S - MPI_Cart_create {} {} {} {} {} {}",
+                                      commOld,
+                                      ndims,
+                                      dims,
+                                      periods,
+                                      reorder,
+                                      newCommPtr);
+
+    ContextWrapper ctx(commOld);
+
+    // Set up the communicator object in the WASM memory. Note that we return
+    // a pointer to it.
+    I32 tmpPtr = Runtime::memoryRef<I32>(ctx.memory, newCommPtr);
+    faabric_communicator_t* newComm =
+      &Runtime::memoryRef<faabric_communicator_t>(ctx.memory, tmpPtr);
+
+    // Cast the original communicator
+    auto origComm =
+      Runtime::memoryRef<faabric_communicator_t>(ctx.memory, commOld);
+
+    // Copy the values from the old the the new
+    *newComm = origComm;
+
+    return MPI_SUCCESS;
+}
+
+/**
+ * Determines process rank in communicator given Cartesian location.
+ */
+WAVM_DEFINE_INTRINSIC_FUNCTION(env,
+                               "MPI_Cart_rank",
+                               I32,
+                               MPI_Cart_rank,
+                               I32 comm,
+                               I32 coords,
+                               I32 rankPtr)
+{
+    faabric::util::getLogger()->debug(
+      "S - MPI_Cart_rank {} {} {}", comm, coords, rankPtr);
+
+    ContextWrapper ctx(comm);
+
+    int* coordsArray = Runtime::memoryArrayPtr<int>(
+      ctx.memory, (Uptr)coords, MPI_CART_MAX_DIMENSIONS);
+    int sideLength =
+      static_cast<int>(std::floor(std::sqrt(ctx.world.getSize())));
+    int rank = sideLength * coordsArray[0] + coordsArray[1];
+    ctx.writeMpiResult<int>(rankPtr, rank);
+
+    return MPI_SUCCESS;
+}
+
+/**
+ * Retrieves the Cartesian topology information associated with a
+ * communicator.
+ *
+ * MPI Topologies are pointless in a serverless environment. Therefore
+ * we return default values (2dim grid) basing on the current world size.
+ *
+ * In particular we define a 2-dim grid with as many processors, leaving
+ * the rest as MPI_UNDEFINED.
+ */
+WAVM_DEFINE_INTRINSIC_FUNCTION(env,
+                               "MPI_Cart_get",
+                               I32,
+                               MPI_Cart_get,
+                               I32 comm,
+                               I32 maxdims,
+                               I32 dims,
+                               I32 periods,
+                               I32 coords)
+{
+    faabric::util::getLogger()->debug(
+      "S - MPI_Cart_get {} {} {} {} {}", comm, maxdims, dims, periods, coords);
+
+    ContextWrapper ctx(comm);
+
+    auto commPtr =
+      &Runtime::memoryRef<faabric_communicator_t>(ctx.memory, comm);
+    faabric::util::getLogger()->debug(
+      "Cart_get w/ comm: {}. Id: {}", comm, commPtr->id);
```

Comment:
This sort of print debugging should be removed (there should be a test to check this works)

Reply:
I totally missed this one. Thanks